infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/AsyncStorageProvisioner.java

```diff
@@ -0,0 +1,370 @@
+      @Named("che.infra.kubernetes.pvc.name") String configuredPVCName,
+      @Named("che.infra.kubernetes.pvc.storage_class_name") String pvcStorageClassName,
+      SshManager sshManager,
+      OpenShiftClientFactory openShiftClientFactory) {
+    this.pvcQuantity = pvcQuantity;
+    this.storageImage = image;
+    this.accessMode = accessMode;
+    this.strategy = strategy;
+    this.configuredPVCName = configuredPVCName;
+    this.pvcStorageClassName = pvcStorageClassName;
+    this.sshManager = sshManager;
+    this.clientFactory = openShiftClientFactory;
+  }
+
+  public void provision(OpenShiftEnvironment osEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+    if (!"true".equals(osEnv.getAttributes().get(ASYNC_PERSIST_ATTRIBUTE))) {
+      return;
+    }
+
+    if ("true".equals(osEnv.getAttributes().get(ASYNC_PERSIST_ATTRIBUTE))
+        && !"common".equals(strategy)) {
+      String message =
+          format(
+              "Workspace configuration not valid: Asynchronous storage available only for 'common' PVC strategy, but got %s",
+              strategy);
+      LOG.warn(message);
+      osEnv.addWarning(new WarningImpl(4200, message));
+      throw new InfrastructureException(message);
+    }
+
+    if ("true".equals(osEnv.getAttributes().get(ASYNC_PERSIST_ATTRIBUTE))
+        && !isEphemeral(osEnv.getAttributes())) {
+      String message =
+          "Workspace configuration not valid: Asynchronous storage available only if attribute 'persistVolumes' set to false";
+      LOG.warn(message);
+      osEnv.addWarning(new WarningImpl(4200, message));
+      throw new InfrastructureException(message);
+    }
+
+    String namespace = identity.getInfrastructureNamespace();
+    KubernetesClient oc = clientFactory.create(identity.getWorkspaceId());
+
+    boolean isPvcExist =
+        oc.persistentVolumeClaims()
+            .inNamespace(namespace)
+            .list()
+            .getItems()
+            .stream()
+            .anyMatch(
+                (Predicate<PersistentVolumeClaim>)
+                    pvc -> pvc.getMetadata().getName().equals(configuredPVCName));
+
+    if (!isPvcExist) {
+      PersistentVolumeClaim pvc =
+          KubernetesObjectUtil.newPVC(
+              configuredPVCName, accessMode, pvcQuantity, pvcStorageClassName);
+      oc.persistentVolumeClaims().inNamespace(namespace).create(pvc);
+    }
+
+    String configMapName = namespace + ASYNC_STORAGE_CONFIG;
+    boolean isMapExist =
+        oc.configMaps()
+            .inNamespace(namespace)
+            .list()
+            .getItems()
+            .stream()
+            .anyMatch(
+                (Predicate<ConfigMap>)
+                    configMap -> configMap.getMetadata().getName().equals(configMapName));
+
+    if (!isMapExist) {
+      ConfigMap configMap = createConfigMap(namespace, configMapName, identity, osEnv);
+      oc.configMaps().inNamespace(namespace).create(configMap);
```

Comment:
Please consider using `KubernetesConfigsMaps` class (& others too) for object creation, they provide proper objects labeling (with workspaceId) and accepts `clientFactory` so you don't have to maintain own client.

Reply:
Thanks, I don't know about this class

Reply:
@mshaposhnik @metlos Looks like [KubernetesConfigsMaps](https://github.com/eclipse/che/blob/master/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesConfigsMaps.java#L34) has `package-private` access. Can I make it public?

Reply:
Don't see a reason why it shouldn't be public.

Reply:
The main point of package private K8sConfigMaps constructor - it's not supposed to be used directly but through K8sNamespace. So, instead of creating new instance of K8sConfigMaps I would investigate if it makes sense to create  K8sNamespace here...

Reply:
ok, guys thanks for your suggest but probably I will keep it as is. I will investigate and improve it later in next sprints